doc: zbus: update documentation with changes for 3.5

rodrigopex opened the pull request:
Add documentation for the new way to allocate observers, the message subscribers, and the confirmed message sample. And 
 add the breaking changes to the migration guide.

Structural changes: 

* Replace Zbus with ZBus;
* Remove the message from the slogan. Zephyr Bus instead of Zephyr Message Bus;
* Change the images.

rodrigopex committed:
doc: zbus: update documentation with changes for 3.5

Add documentation for the new way to storage observers, the message
subscribers, and the confirmed message sample.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex commented:
@kartben, I would like to share the Figma file I used to draw the bus diagrams with people. The files are here: https://www.figma.com/community/file/1292866458780627559. What do you think is the best way to do that?

kartben commented on doc/services/zbus/index.rst:
```diff
@@ -1,58 +1,74 @@
 
-Zephyr message bus (zbus)
-#########################
+Zephyr Bus (ZBus)
```
If you want to spell it ZBus from now on please also consider updating other places where it might be referred to as "zbus", for example in code samples. FWIW I liked the lower-case naming :)

kartben commented on doc/services/zbus/index.rst:
```diff
@@ -1,58 +1,74 @@
 ********
-
-Threads can broadcast messages to all interested observers using zbus. Many-to-many communication is possible. The bus implements message-passing and publish/subscribe communication paradigms that enable threads to communicate synchronously or asynchronously through shared memory. The communication through zbus is channel-based, where threads publish and read to and from using messages. Additiona ...
+Threads can broadcast messages to all interested observers using ZBus. Many-to-many communication is possible. The bus implements message-passing and publish/subscribe communication paradigms that enable threads to communicate synchronously or asynchronously through shared memory. The communication through ZBus is channel-based, where threads publish and read to and from using messages. Additiona ...
```
documentation should also following coding guidelines and have lines wrap at 100 characters. 
The long lines are making the actuall differences hard to review :/

rodrigopex replied:
Do you suggest any formatting tool like `clang-format` to do this? I am trying to use the [rstfmt](https://pypi.org/project/rstfmt/), but it is not working because of the sphinx commands. I suggest standardizing that like we do with the code. I will try out something here to improve the documentation format and organization.

rodrigopex replied:
The old and good Neovim (Lazyvim, in my case), with its `gw` keybinding, immensely helped.

rodrigopex replied:
The only issue here is that the new line-split file will also be hard to keep track of because it will change all the text.

kartben replied:
Just make a separate commit for the formatting-only changes?

rodrigopex replied:
Ok. I will do.

rodrigopex replied:
It will not be possible. There are tons of rework to do. I tried to do that, but there were many conflicts and other issues. I am rewriting a significant portion of that. It may take more time to review, but it would be better to check all to see if it is still consistent. Sorry for the burden.

kartben commented on doc/services/zbus/index.rst:
```diff
@@ -1,58 +1,74 @@
 ********
-
-Threads can broadcast messages to all interested observers using zbus. Many-to-many communication is possible. The bus implements message-passing and publish/subscribe communication paradigms that enable threads to communicate synchronously or asynchronously through shared memory. The communication through zbus is channel-based, where threads publish and read to and from using messages. Additiona ...
+Threads can broadcast messages to all interested observers using ZBus. Many-to-many communication is possible. The bus implements message-passing and publish/subscribe communication paradigms that enable threads to communicate synchronously or asynchronously through shared memory. The communication through ZBus is channel-based, where threads publish and read to and from using messages. Additiona ...
```
I would also consider making this several paragraphs, as it is a bit intimidating to have the introduction to ZBus/zbus be such a big block of text

<img width="660" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/b65bcf3a-68d8-4936-96e3-ec3211083a00">

kartben reviewed (commented):
added a few comments. Anything you could do to make the diff easier to review would be appreciated!

kartben commented:
> @kartben, I would like to share the Figma file I used to draw the bus diagrams with people. The files are here: https://www.figma.com/community/file/1292866458780627559. What do you think is the best way to do that?

I would add a comment at the top of the index.rst

```
..
   Note to documentation authors: the diagrams included in this documentation page were designed
   using the following Figma library: xxx. 
```

rodrigopex reviewed (commented):
Migration guide entries added.

jukkar commented on doc/releases/migration-guide-3.5.rst:
```diff
@@ -235,6 +235,16 @@ Required changes
   This means that an extra linker stage is no longer necessary if this option is
   not enabled.
 
+* ZBus runtime observers implementation now relies on the HEAP memory instead of a memory slab. The
+  configuration changed. So, to keep your runtime observers code working correctly, you need to:
```
The sentence
```
The configuration changed.
```
looks quite lonely. Is there something missing from it or perhaps more explanation could be added to it?

rodrigopex replied:
For sure. I will adjust that.

rodrigopex replied:
Done!

rodrigopex reviewed (commented):
Improve migration guide sentence regarding runtime observers.

rodrigopex committed:
doc: zbus: add entries to the migration guide

ZBus had one break change related to the runtime observers'
configuration, and the VDED delivery sequence has changed. This commit
adds entries on the migration guide about the mentioned changes.

Signed-off-by: Rodrigo Peixoto <rodrigopex@gmail.com>

rodrigopex commented:
@jukkar and @jhedberg, could you please take a look at that again? There was some conflict, and I needed to submit that again after accidentally accepting a merge offered by GitHub.

kartben commented on doc/releases/migration-guide-3.5.rst:
```diff
@@ -271,6 +271,17 @@ Required changes
+  Thus, zbus' configuration (kconfig) related to runtime observers has changed. To keep your runtime
+  observers code working correctly, you need to:
+
+  - Replace the integer :kconfig:option:`CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE` with the boolean
```
nit: it probably doesn't make sense to use :kconfig: to point to an option that doesn't exist anymore

```suggestion
  - Replace the integer ``CONFIG_ZBUS_RUNTIME_OBSERVERS_POOL_SIZE`` with the boolean
```

kartben commented on doc/services/zbus/index.rst:
```diff
@@ -23,145 +41,331 @@ Threads can broadcast messages to all interested observers using zbus. Many-to-m
+.. warning::
+   ZBus uses :zephyr_file:`include/zephyr/net/buf.h` (network buffers) to exchange data with message
+   subscribers. So, chose carefully the configurations
+   :c:macro:`CONFIG_ZBUS_MSG_SUBSCRIBER_NET_BUF_POOL_SIZE` and :c:macro:`CONFIG_HEAP_MEM_POOL_SIZE`.
```
```suggestion
   :kconfig:`CONFIG_ZBUS_MSG_SUBSCRIBER_NET_BUF_POOL_SIZE` and :kconfig:`CONFIG_HEAP_MEM_POOL_SIZE`.
```

kartben commented on doc/services/zbus/index.rst:
```diff
@@ -212,13 +417,44 @@ The following code defines and initializes a regular channel and its dependencie
 
 
-Channels can have a ``validator function`` that enables a channel to accept only valid messages. Publish attempts invalidated by hard channels will return immediately with an error code. This allows original creators of a channel to exert some authority over other developers/publishers who may want to piggy-back on their channels. The following code defines and initializes a :dfn:`hard channel` a ...
+Channels can have a ``validator function`` that enables a channel to accept only valid messages.
```
I am confused about the use of "code" formatting for `validator function`, is this on purpose? I think you probably meant to use single backquotes?

kartben commented on doc/services/zbus/index.rst:
```diff
@@ -212,13 +417,44 @@ The following code defines and initializes a regular channel and its dependencie
+allows original creators of a channel to exert some authority over other developers/publishers who
+may want to piggy-back on their channels. The following code defines and initializes a :dfn:`hard
+channel` and its dependencies. Only valid messages can be published to a :dfn:`hard channel`. It is
+possible because a ``validator function`` was passed to the channel's definition. In this example,
```
ditto

kartben commented on doc/services/zbus/index.rst:
```diff
@@ -482,23 +765,40 @@ It is possible to add observers to channels in runtime. This feature uses the he
-* :zephyr:code-sample:`zbus-dyn-channel` demonstrates how to use dynamically allocated exchanging data in zbus;
-* :zephyr:code-sample:`zbus-uart-bridge` shows an example of sending the operation of the channel to a host via serial;
-* :zephyr:code-sample:`zbus-remote-mock` illustrates how to implement an external mock (on the host) to send and receive messages to and from the bus.
-* :zephyr:code-sample:`zbus-runtime-obs-registration` illustrates a way of using the runtime observer registration feature;
+* :zephyr:code-sample:`zbus-work-queue` shows how to define and use different kinds of observers.
+  Note there is an example of using a work queue instead of executing the listener as an execution
+  option;
+* :zephyr:code-sample:`zbus-msg-subscriber` illustrates how to use message subscribers.
+* :zephyr:code-sample:`zbus-dyn-channel` demonstrates how to use dynamically allocated exchanging
+  data in zbus;
+* :zephyr:code-sample:`zbus-uart-bridge` shows an example of sending the operation of the channel to
+  a host via serial;
+* :zephyr:code-sample:`zbus-remote-mock` illustrates how to implement an external mock (on the host)
+  to send and receive messages to and from the bus.
+* :zephyr:code-sample:`zbus-runtime-obs-registration` illustrates a way of using the runtime
+  observer registration feature;
+* :zephyr:code-sample:`zbus-confirmed-channel` implements a way of implement confirmed channel only
+  with subscribers.
 * :zephyr:code-sample:`zbus-benchmark` implements a benchmark with different combinations of inputs.
```
btw it makes me realize that it would maybe make sense to provide a Sphinx directive that outputs the equivalent of what shows up in the API reference, as this could be useful to list samples anywhere.

<img width="643" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/cbb57472-7692-41ca-af90-c4d7ec40289b">

rodrigopex replied:
@kartben, how could you change the title of the note? 

![image](https://github.com/zephyrproject-rtos/zephyr/assets/35280/909b023a-d98d-4874-9576-30efee113e59)

Should I remove the Samples sections?

kartben replied:
Sorry I should have been clearer, and I think it's fine if you live it as-is for now. I was mostly talking to myself and thinking that I should add to the code-sample sphinx extension a way to easily output the list of samples related to a specific set of doxygengroups, as it might be useful in situations like yours. 

That being said, this section pretty much "paraphrases" the list of samples that also already show up automatically in the API reference, mere lines below, so if you think that list is enough and easy enough to not miss for folks, then maybe just drop your Samples section altogether (but again, also 100% fine with me if you'd like to keep it!)

kartben reviewed (changes requested):
Great stuff @rodrigopex! I based my review mostly on the looks of the final rendered page rather than looking specifically at the diff -- spotted a few formatting style issues.

rodrigopex reviewed (commented):
@kartben, I changed most of your requests. The only thing I still need to change is the samples. I'm just waiting for your feedback to remove that or not. But the rest is ok.
